Disabled including binaries for generators in latest .net sdk preview

<!--
Please keep in mind that the GitHub issue tracker is not intended as a general support forum, but for reporting **non-security** bugs and feature requests.

If you believe you have an issue that affects the SECURITY of the platform, please do NOT create an issue and instead email your issue details to secure@microsoft.com. Your report may be eligible for our [bug bounty](https://www.microsoft.com/en-us/msrc/bounty-dot-net-core) but ONLY if it is reported through email.
For other types of questions, consider using [StackOverflow](https://stackoverflow.com).

-->

### Describe the bug
When trying to include a binary file for use in a generator, the following error appears:
```
error CS2015: is a binary file instead of a text file
```
The peculiarity is that this appears only at the first attempt to assemble.
There is a bug here either in that including binaries for generators was disabled in this preview version, or in that it only works for the first build attempt.
In the second case, there is a question - why was the use of binary files prohibited? I have a library that generates code to conveniently use EmbeddedResources based on binary AdditionalFiles, I use it in tests - https://github.com/HavenDV/H.Resources.Generator

### To Reproduce
Include any binary file:
```xml
<ItemGroup>
  <AdditionalFiles Include="*.*" />
</ItemGroup>
```

### Exceptions (if any)
<!-- 
Include the exception you get when facing this issue
-->

### Further technical details
<img width="395" alt="image" src="https://user-images.githubusercontent.com/3002068/151090977-1d47d582-80ab-4ee6-8fc0-aed706ef0b45.png">
VS 17.1 preview 4.0

--- a/src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs
+++ b/src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs
@@ -32,23 +32,22 @@ namespace Microsoft.CodeAnalysis
                 return false;
             }
 
-            try
-            {
-                var xText = GetTextOrNullIfBinary(x);
-                var yText = GetTextOrNullIfBinary(y);
+            var xText = GetTextOrNullIfBinary(x);
+            var yText = GetTextOrNullIfBinary(y);
 
-                if (xText is null || yText is null || xText.Length != yText.Length)
-                {
-                    return false;
-                }
-
-                return ByteSequenceComparer.Equals(xText.GetChecksum(), yText.GetChecksum());
-            }
-            catch (InvalidDataException)
+            // If xText and yText are both null, then the additional text is observably not changed
+            // and can be treated as equal.
+            if (xText is null && yText is null)
             {
-                // Either x, y, or both
-                throw;
+                return true;
             }
+
+            if (xText.Length != yText.Length)
+            {
+                return false;
+            }
+
+            return ByteSequenceComparer.Equals(xText.GetChecksum(), yText.GetChecksum());
         }
 
         public int GetHashCode(AdditionalText obj)

--- a/src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs
+++ b/src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs
@@ -2,11 +2,11 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.Diagnostics.CodeAnalysis;
+using System.IO;
 using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis
@@ -32,21 +32,42 @@ namespace Microsoft.CodeAnalysis
                 return false;
             }
 
-            var xText = x.GetText();
-            var yText = y.GetText();
-
-            if (xText is null || yText is null || xText.Length != yText.Length)
+            try
             {
-                return false;
-            }
+                var xText = GetTextOrNullIfBinary(x);
+                var yText = GetTextOrNullIfBinary(y);
 
-            return ByteSequenceComparer.Equals(xText.GetChecksum(), yText.GetChecksum());
+                if (xText is null || yText is null || xText.Length != yText.Length)
+                {
+                    return false;
+                }
+
+                return ByteSequenceComparer.Equals(xText.GetChecksum(), yText.GetChecksum());
+            }
+            catch (InvalidDataException)
+            {
+                // Either x, y, or both
+                throw;
+            }
         }
 
         public int GetHashCode(AdditionalText obj)
         {
             return Hash.Combine(PathUtilities.Comparer.GetHashCode(obj.Path),
                                 ByteSequenceComparer.GetHashCode(obj.GetText()?.GetChecksum() ?? ImmutableArray<byte>.Empty));
+        }
+
+        private static SourceText? GetTextOrNullIfBinary(AdditionalText text)
+        {
+            try
+            {
+                return text.GetText();
+            }
+            catch (InvalidDataException)
+            {
+                // InvalidDataException is thrown when the underlying text is binary
+                return null;
+            }
         }
     }
 }

--- a/src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs
+++ b/src/Compilers/Core/Portable/InternalUtilities/AdditionalTextComparer.cs
@@ -42,7 +42,7 @@ namespace Microsoft.CodeAnalysis
                 return true;
             }
 
-            if (xText.Length != yText.Length)
+            if (xText is null || yText is null || xText.Length != yText.Length)
             {
                 return false;
             }

